### PR TITLE
Drivers: SPI & I2C config simplifications (3)

### PR DESCRIPTION
**Pull request by bkueng**
This is the last one in the series of https://github.com/PX4/Firmware/pull/14386: it updates the remaining drivers and removes the old board configurations.

Other changes:
- simplified the chip-select code, by avoiding deselecting all others. This reduces CPU load by almost 1% on a Pixhawk cube (bringing us back to where it was before). @davids5 can you have a look? It's in https://github.com/PX4/Firmware/commit/233479db71a14b3aa85c271e974ab6eed42cefde.
- disable SPI locking, if the configuration allows it (i.e. if no NuttX drivers run on the same bus). Reduces CPU load on a Pixhawk cube by **4%**, on pixhawk 4 and v5x by **2%**.

### Testing
Tested on:
- pixhawk 4
- v5x
- pixracer
- pixhawk cube
- nxp/fmuk66-v3

This is [WIP], because there are 3 more drivers remaining that would create conflicts:
- fxas21002c, fxos8701cq (https://github.com/PX4/Firmware/pull/14168)
- ~~QMC5883 (https://github.com/PX4/Firmware/pull/14384). @dagar, I can do it on top of that PR, but I don't have HW to test.~~ https://github.com/PX4/Firmware/pull/14438#issuecomment-601800588

### CODE REVIEW

**Comment by dagar** [comment]
> QMC5883 (#14384). @dagar, I can do it on top of that PR, but I don't have HW to test.

I would go ahead as is, I wasn't going to deal with it until after this release. I'd like to get the sensor rotations to a sane consistent state, including fully understanding what's currently in the wild for external mags.

File: src/drivers/imu/invensense/icm20602/ICM20602.cpp [hunk 1]
```diff
@@ -392,49 +381,21 @@ void ICM20602::DataReady()
...
-#elif defined(GPIO_DRDY_ICM_2060X)
-	ret_setevent = px4_arch_gpiosetevent(GPIO_DRDY_ICM_2060X, false, false, false, nullptr, nullptr);
-#endif
-
-	return (ret_setevent == 0);
+	if (_drdy_gpio == 0) {
+		return false;
+	}
+
+	return px4_arch_gpiosetevent(_drdy_gpio, false, false, false, nullptr, nullptr) == 0;
```
Comment (dagar) [inline src/drivers/imu/invensense/icm20602/ICM20602.cpp]: Beautiful!

File: src/drivers/imu/invensense/icm20602/ICM20602.hpp [hunk 2]
```diff
@@ -143,11 +152,9 @@ class ICM20602 : public device::SPI, public px4::ScheduledWorkItem
 		WAIT_FOR_RESET,
 		CONFIGURE,
 		FIFO_READ,
-		REQUEST_STOP,
-		STOPPED,
 	};
 
-	px4::atomic<STATE> _state{STATE::RESET};
+	STATE _state{STATE::RESET};
```
Comment (dagar) [inline src/drivers/imu/invensense/icm20602/ICM20602.hpp]: Thanks for cleaning this up. At one point I was experimenting with initiating certain state transitions from the interrupt for NuttX SPI_TRIGGER.

File: boards/nxp/fmuk66-v3/src/spi.cpp [hunk 3]
```diff
@@ -211,7 +211,7 @@ __EXPORT int fmuk66_spi_bus_initialize(void)
 
 	/* Configure SPI-based devices */
 
-	struct spi_dev_s *spi_sensors = px4_spibus_initialize(1);
+	struct spi_dev_s *spi_sensors = px4_spibus_initialize(PX4_BUS_NUMBER_TO_PX4(1));
```
Comment (dagar) [inline boards/nxp/fmuk66-v3/src/spi.cpp]: Do we actually need to map the bus number at all? In the board (i2c/spi) we're adding an offset (PX4_BUS_NUMBER_FROM_PX4), then in the micro_hal we're removing it (PX4_BUS_NUMBER_FROM_PX4).

Where would it break down if we simply used the real underlying bus number end-to-end? It's not really user facing (no external port numbering), and in the few cases it kind of is (decoding device ids?) it adds even more confusion.

What am I missing?

File: boards/nxp/fmuk66-v3/src/spi.cpp [hunk 4]
```diff
@@ -131,6 +131,7 @@ __EXPORT void board_spi_reset(int ms, int bus_mask)
 			for (int i = 0; i < SPI_BUS_MAX_DEVICES; ++i) {
 				if (px4_spi_buses[bus].devices[i].cs_gpio != 0) {
 					kinetis_pinconfig(px4_spi_buses[bus].devices[i].cs_gpio);
+					kinetis_gpiowrite(px4_spi_buses[bus].devices[i].cs_gpio, 1);
```
Comment (davids5) [inline boards/nxp/fmuk66-v3/src/spi.cpp]: The default of of a CS GPIO had better have GPIO_OUTPUT_SET (GPIO_OUTPUT_ONE) so Why is this here?

File: boards/nxp/fmuk66-v3/src/spi.cpp (hunk 3 above)
Comment (davids5) [reply boards/nxp/fmuk66-v3/src/spi.cpp]: The mapping was done to normalize to the STM32 way of counting busses 1-N.  We can change it but does it ripple to `-b` parameter somewhere? If it does common cod will not work.

File: boards/nxp/fmurt1062-v1/src/spi.cpp [hunk 5]
```diff
@@ -372,6 +361,7 @@ __EXPORT void board_spi_reset(int ms, int bus_mask)
 			for (int i = 0; i < SPI_BUS_MAX_DEVICES; ++i) {
 				if (px4_spi_buses[bus].devices[i].cs_gpio != 0) {
 					imxrt_config_gpio(px4_spi_buses[bus].devices[i].cs_gpio);
+					imxrt_gpio_write(px4_spi_buses[bus].devices[i].cs_gpio, 1);
```
Comment (davids5) [inline boards/nxp/fmurt1062-v1/src/spi.cpp]: "

File: boards/px4/fmu-v2/src/spi.cpp [hunk 6]
```diff
@@ -35,6 +35,116 @@
 #include <px4_arch/spi_hw_description.h>
 #include <drivers/drv_sensor.h>
 
+/* Due to inconsistent use of chip select and dry signal on
```
Comment (davids5) [inline boards/px4/fmu-v2/src/spi.cpp]:
```suggestion
/* Due to inconsistent use of chip select and data ready signals on
```

File: boards/px4/fmu-v3/src/spi.cpp [hunk 7]
```diff
@@ -35,6 +35,116 @@
 #include <px4_arch/spi_hw_description.h>
 #include <drivers/drv_sensor.h>
 
+/* Due to inconsistent use of chip select and dry signal on
```
Comment (davids5) [inline boards/px4/fmu-v3/src/spi.cpp]:
```suggestion
/* Due to inconsistent use of chip select and data ready signal on
```

**Comment by dagar** [comment]
I rebased on master and resolved the conflicts (primarily with intel aero adc).

File: boards/nxp/fmuk66-v3/src/spi.cpp (hunk 3 above)
Comment (dagar) [reply boards/nxp/fmuk66-v3/src/spi.cpp]: There isn't common code like that, it's all board specific and I would argue it makes the standalone `-b` usage far more confusing.

**Comment by dagar** [comment]
> How does the code size vs master look?

@davids5 At the moment I'm seeing +248 bytes on fmu-v2.

File: boards/nxp/fmuk66-v3/src/spi.cpp (hunk 3 above)
Comment (bkueng) [reply boards/nxp/fmuk66-v3/src/spi.cpp]: For I2C it is used as an array index. I didn't want to go through that within these changes, so I left the offset in for now.
I'm all for removing it as well.

File: boards/px4/fmu-v2/src/spi.cpp (hunk 6 above)
Comment (bkueng) [reply boards/px4/fmu-v2/src/spi.cpp]: Fixed.

File: boards/px4/fmu-v3/src/spi.cpp (hunk 7 above)
Comment (bkueng) [reply boards/px4/fmu-v3/src/spi.cpp]: Fixed.

File: boards/nxp/fmuk66-v3/src/spi.cpp (hunk 4 above)
Comment (bkueng) [reply boards/nxp/fmuk66-v3/src/spi.cpp]: I was being paranoid :) Indeed, it is, and I removed these.

**Comment by dagar** [comment]
Overall increase on px4_fmu-v2_default is minimal (a couple hundred bytes), although on other platforms it's quite substantial (> 10 KB increase on px4_fmu-v4_default). At a glance this is probably due to some of the optional drivers not being GC'd at link time due to missing SPIDEV ifdefs.

**Comment by davids5** [comment]
@bkueng @dagar - I have run it. 

So for SPI it is clear at to what device is where located on what bus.
Where is that mapping for I2C? Is is just in rc.board_sensors? 

Also I found that the help was confusing.

` [-b <val>]  bus (board-specific internal (default=all) or n-th external (default=1)`

If I have a ist8130 on I2C1 and I2C2 what is the command to start each one separately?

**Comment by bkueng** [comment]
> fmu-v4pro missing an IMU

I'm seeing this as well:
```
WARN  [mpu9250] AK8963: bad id 171 retries 9
WARN  [mpu9250] AK8963: bad id 171 retries 8
WARN  [mpu9250] AK8963: bad id 171 retries 7
WARN  [mpu9250] AK8963: bad id 171 retries 6
WARN  [mpu9250] AK8963: bad id 171 retries 5
WARN  [mpu9250] AK8963: bad id 171 retries 4
WARN  [mpu9250] AK8963: bad id 171 retries 3
WARN  [mpu9250] AK8963: bad id 171 retries 2
WARN  [mpu9250] AK8963: bad id 171 retries 1
WARN  [mpu9250] AK8963: bad id 171 retries 0
ERROR [mpu9250] AK8963: failed to initialize, disabled!
```
Adding a `sleep 1` in front fixes it, so might be a power-up issue?

@davids5 
> Where is that mapping for I2C? Is is just in rc.board_sensors?
Also I found that the help was confusing.
[-b <val>] bus (board-specific internal (default=all) or n-th external (default=1)
If I have a ist8130 on I2C1 and I2C2 what is the command to start each one separately?

The logic is the same for I2C and SPI: `-X -b <i>` gives you the i-th external bus (should follow the label on the board), while `-I -b <i>` gives you the i-th bus of the board (for the internal ones).

**Comment by dagar** [comment]
I'll follow up with the mpu9250 separately before the release (https://github.com/PX4/Firmware/pull/14441, etc).

**Comment by dagar** [comment]
> Where is that mapping for I2C? Is is just in rc.board_sensors?

Let's discuss this as a follow up. For the internal I2C I'd like to see if we can get to simple hierarchical representation (per bus) with optional address overrides.

**Comment by dagar** [comment]
> If I have a ist8130 on I2C1 and I2C2 what is the command to start each one separately?

@davids5 you can still do it manually (v5x ina226 example).
https://github.com/PX4/Firmware/blob/f797eae0d02aaf0bd13695eb1adbd6c36e87b002/boards/px4/fmu-v5x/init/rc.board_sensors#L8-L10

If you don't set the specific bus and just set `-X` for external I2C it will iterate through all externals.

**Comment by dagar** [comment]
This is a huge improvement in consistency for driver interaction. Great work Beat!
